src/Stmt.h

```diff
@@ -122,6 +122,7 @@ class IfStmt final : public ExprStmt {
     StmtPtr DoReduce(Reducer* c) override;
 
     bool NoFlowAfter(bool ignore_break) const override;
+    bool CouldReturn(bool ignore_break) const override;
```

Comment:
This could've also been realized as a `TraversalCallback`, no? (and probably `NoFlowAfter`, too).

Not a too strong opinion, but I'm mostly looking at things from the perspective of "what if someone wanted to do this out-of-tree, do we provide the right tools" and "we should use these tools in-tree as well".

Reply:
I can see some of these going either way. If you look at the `TraversalCallback`s used for script optimization, they generally have (1) `switch` statements to determine "which node is it this time", along with (2) logic that implements analysis that a node can't readily do locally for itself. For `CouldReturn`, using traversal will (a) do a whole bunch of unnecessary calls, (b) have some complexity due to needing to track potentially nested `switch` nodes (which change the `ignore_break` setting). In this case, I also wanted to leave room for nodes caching the result, which the traversal framework doesn't cleanly support since it's only for `const` objects.